### PR TITLE
Utility to build voltage pulse function (piece-wise linear).

### PR DESCRIPTION
**Pull request by roelof-groenewald**
Functionality was added to construct the expression (string) that will enable voltage pulses to be applied to electrodes.

### CODE REVIEW

File: mewarpx/mewarpx/utils_store/pulsing.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          ____                               ____
+         /    \                             /    \
+    ____/      \___________________________/      \_____
+
+    Arguments:
+        V_off (float): Electrode bias in the off-phase in Volt
+        V_on (float): Electrode voltage at the peak of the pulse in Volt
+        pulse_period (float): Full period from pulse start to pulse end in
+            seconds
+        pulse_length (float): Time duration of on-pulse phase in seconds
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/utils_store/pulsing.py]: Clarify if this includes t_rise and t_fall (I assume not, but being explicit is good).

File: mewarpx/mewarpx/utils_store/pulsing.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    pulse with form
+          ____                               ____
+         /    \                             /    \
+    ____/      \___________________________/      \_____
+
+    Arguments:
+        V_off (float): Electrode bias in the off-phase in Volt
+        V_on (float): Electrode voltage at the peak of the pulse in Volt
+        pulse_period (float): Full period from pulse start to pulse end in
+            seconds
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/utils_store/pulsing.py]:
```suggestion
        pulse_period (float): Full period from pulse start to the subsequent
            pulse start in seconds
```

File: mewarpx/tests/test_utils_pulsing.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
+"""Test pseudo 1D diode run with the superLU solver."""
```
Comment (peterscherpelz) [inline mewarpx/tests/test_utils_pulsing.py]: Modify comment

File: mewarpx/tests/test_utils_pulsing.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+from mewarpx.utils_store import util as mwxutil, pulsing
+
+def test_utils_pulsing_expr():
+
+    pulse_expr = pulsing.linear_pulse_function(
+        V_off=-1.0, V_on=7.5, pulse_period=20e-9, pulse_length=10e-9,
+        t_rise=5e-9, t_fall=5e-9
+    )
+    print(pulse_expr)
+    assert pulse_expr == "modT=(t-5e-09); p1=(modT>=0 and modT<5e-09); p2=(modT>=5e-09 and modT<(5e-09+1e-08)); p3=(modT>=(5e-09+1e-08) and modT<(5e-09+1e-08+5e-09)); p4=if(p1 or p2 or p3, 0, 1); dV=7.5--1.0; p1*(-1.0+dV/5e-09*modT)+p2*7.5+p3*(7.5-dV/5e-09*(modT-5e-09-1e-08))+p4*-1.0"
```
Comment (peterscherpelz) [inline mewarpx/tests/test_utils_pulsing.py]: I'm honestly not convinced by the utility of this as it mostly tests if it's implemented exactly the same, not that it's functionally the same. The test below seems to test functionality better. But I don't have a reason to remove it.

Comment (peterscherpelz) [reply mewarpx/tests/test_utils_pulsing.py]: I guess a reason to leave it is that if it passes and the one below doesn't, we know that something in getvoltage() or some other mewarpx/WarpX thing is causing that run to fail, not this expression creator. So I'm fine leaving it.

**Review by peterscherpelz — COMMENTED** [review]
No major concerns! Are you planning to add plotting functionality here or make it a separate PR?

File: mewarpx/tests/test_utils_pulsing.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        return voltages
+
+    pulse_function = type('testclass', (object,),
+        {
+            'get_voltage': get_voltage,
+            'V_off': V_off, 'V_on': V_on, 'pulse_period': pulse_period,
+            'pulse_length': pulse_length, 't_rise': t_rise,
+            't_fall': t_fall, 'wait_time': wait_time
+        }
+    )()
```
Comment (peterscherpelz) [inline mewarpx/tests/test_utils_pulsing.py]: It makes more sense to me to pull all this out into a separate class in `pulsing.py`. You could even make `plot_pulse` and `linear_pulse_function` member functions at that point. But having the plotting interface easily accessible by the user makes more sense to me than this syntax...

File: mewarpx/mewarpx/utils_store/pulsing.py [hunk 6]
```diff
@@ -3,10 +3,20 @@
...
 from mewarpx.mwxrun import mwxrun
 from pywarpx import callbacks
 
+
+# namedtuple is a convenient way of making a class with given properties
+PulseFunction = namedtuple(
+    'PulseFunction',
+    ['get_voltage', 'V_off', 'V_on', 'pulse_period', 'pulse_length',
+     't_rise', 't_fall', 'wait_time']
+)
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/utils_store/pulsing.py:17]: Nice! Yeah this is definitely a much cleaner implementation. Thanks!

**Pull request merged**